oblib/parser.py

```diff
@@ -0,0 +1,216 @@
+            # if "id" in fact:
+            #     kwargs["id"] = fact["id"]
+
+            if "xbrl:periodStart" in fact["aspects"] and "xbrl:periodEnd" in fact["aspects"]:
+                if fact["aspects"]["xbrl:periodStart"] == fact["aspects"]["xbrl:periodEnd"]:
+                    kwargs["instant"] = util.convert_json_datetime(fact["aspects"]["xbrl:periodStart"])
+                else:
+                    kwargs["duration"] = {}
+                    kwargs["duration"]["start"] = util.convert_json_datetime(fact["aspects"]["xbrl:periodStart"])
+                    kwargs["duration"]["end"] = util.convert_json_datetime(fact["aspects"]["xbrl:periodEnd"])
+            else:
+                kwargs["duration"] = "forever"
+
+            if "xbrl:unit" in fact["aspects"]:
+                kwargs["unit_name"] = fact["aspects"]["xbrl:unit"]
+
+            # Add axis to kwargs if item is an axis.
+            for axis_chk in fact["aspects"]:
+                if "Axis" in axis_chk:
+                    kwargs[axis_chk.split(":")[1]] = fact["aspects"][axis_chk]
+
+            entrypoint.set(fact["aspects"]["xbrl:concept"], fact["value"], **kwargs)
+
+        return entrypoint
+
+    def from_JSON(self, filename):
+        """ Imports XBRL as JSON from the given filename. """
+
+        infile = open(filename, "r")
+        s = infile.read()
+        infile.close()
+        return self.from_JSON_string(s)
+
+    def from_XML_string(self, xml_string):
+        """ Loads the Entrypoint from an XML string """
+
+        # Create an entrypoint.
+        # TODO: Should not have to initiatilize taxonomy and Entrypoint type should not be hardcoded.
+        entrypoint = data_model.Entrypoint("CutSheet", self._taxonomy)
+
+        # TODO: Supply implementation upon completion of prototype code.
+        return entrypoint
+
+    def from_XML(self, filename):
+        """ Imports XBRL as XML from the given filename. """
+
+        infile = open(filename, "r")
+        s = infile.read()
+        infile.close()
+        self.from_XML_string(s)
+
+    def to_JSON_string(self, entrypoint):
+        """ Returns XBRL as an JSON string given a data model entrypoint. """
+
+        return entrypoint.to_JSON_string()
+
+    def to_JSON(self, entrypoint, filename):
+        """ Exports XBRL as JSON to the given filename given a data model entrypoint. """
+        
+        entrypoint.to_JSON(filename)
+
+    def to_XML_string(self, entrypoint):
+        """ Returns XBRL as an XML string given a data model entrypoint. """
+
+        return entrypoint.to_XML_string()
+
+    def to_XML(self, entrypoint, filename):
+        """ Exports XBRL as XML to the given filename given a data model entrypoint. """
+        
+        entrypoint.to_XML(filename)
+
+    def convert(self, in_filename, out_filename, file_format):
+        """ 
+        Converts and input file (in_filename) to an output file (out_filename) given an input 
+        file format specified by file_format.
```

Comment:
Would help to list the inputs in the docstring.

in_filename: str
    full path to input file
out_filename: str
    full path to output file
file_format: str
    values are "JSON" or "XML"

Reply:
Agreed - we probably need to do this code wide.

Reply:
values are FileFormat['JSON'] or FileFormat['XML'] then?